### PR TITLE
Build a separate `print-api` executable for each supported GHC versions

### PR DESCRIPTION
**Pull request by mmhat**
Then, the `print-api` executable is a simple wrapper that detects the used GHC version and calls the appropriate `print-api-VERSION` exectuable which does the actual work.
The following works (inside the `print-api` source tree):

```bash
# Outputs 9.10.1 in my case
ghc --numeric-version

# Builds and installs print-api and print-api-9.8.2
cabal build print-api:exe:print-api-9.8.2 --with-compiler ghc-9.8.2

# Builds and installs print-api and print-api-9.10.1
cabal build all --with-compiler ghc-9.10.1

# We need to build the project once for each GHC; Otherwise print-api won't find the appropriate packagedb
cabal build all --with-compiler 9.8.2
cabal build all --with-compiler 9.10.1

All of these commands give the correct output using the GHC executable that is "in scope".
print-api -p print-api
ghcup run --ghc 9.8.2 -- print-api -p print-api
ghcup run --ghc 9.10.1 -- print-api -p print-api
```

I also checked in a `install-for-ghcup-compilers.sh` Bash script that installs the executable for all GHCs currently installed by ghcup.
I think this unblocks #13 to some degree, doesn't it?

Also: This tool is awesome! Thanks for writing it!

### CODE REVIEW

**Comment by Kleidukos** [comment]
@mmhat Hi! Thank you so much for your involvement in `print-api`. :) 

Could you please update the release action to produce binaries for the wrapper as well? One per OS/Arch will be enough. :)

**Comment by mmhat** [comment]
@Kleidukos Ok, there is only one issue left with the packaging: The binaries in `print-api-0.1.0.1-Linux-x86_64.tar.gz` release archive should be dynamically linked, but they are not. I couldn't figure out why, and given that I spent much more time today on this issue than I intended, this has to wait...

**Comment by Kleidukos** [comment]
@mmhat Yes sorry, I was confused too at first but never wrote anything down about it in the repository.

**Comment by mmhat** [comment]
@Kleidukos Ok, I think I am done here. Can you review the PR and tell me if there is something left to do for me?

Also, here are some quick remarks (not sure which of those should be an own issue or part of this PR):

1. I think some parts of the CI can be consolidated: Right now we have two workflow containing two jobs. Those four jobs are awfully similar, and we might just merge them (or some of them) into one (or two) jobs/workflows in order to reduce duplication.
    Would such a contribution be accepted?
2. There is some room for UX improvements: For example, we may want to add a `--with-compiler` flag to the wrapper executable to force the usage of a certain GHC compiler.
3. An annoying issue of the current `cabal exec` implementation is that it requires that project was build at least once. See [here](https://github.com/haskell/cabal/issues/6141) for the upstream bug report. Not sure if there is something actionable we can do here; The only thing I can think of is a general suggestion to the user like `Did you run cabal build --with-compiler VERSION ?`.
4. Since the wrapper/worker split of the executables is somewhat fragile we might want a testsuite checking that the `print-api-VERSION` gets picked.
6. The wrapper/worker split of the executables yields some new points in the design space what `print-api` could do if a `print-api-VERSION` executable could not be found. We _could_ extend the scope of it to be a downloader (i.e. download the appropriate bindist if a `print-api-VERSION` is not found) or an installer (i.e. download the source code, build and install the demanded `print-api-VERSION` if it is not found) or print instruction how to obtain it.
7. Also, I think it was nice if the library provides the boiler plate to implement a Cabal testsuite checking API differences.

> @mmhat Yes sorry, I was confused too at first but never wrote anything down about it in the repository.

I left a comment right above the `release` job in the hope that this fact is not forgotten.

**Comment by mmhat** [comment]
Ok, I also updated the documentation :+1:

**Comment by tchoutri** [comment]
> 1. I think some parts of the CI can be consolidated: Right now we have two workflow containing two jobs. Those four jobs are awfully similar, and we might just merge them (or some of them) into one (or two) jobs/workflows in order to reduce duplication.
> Would such a contribution be accepted?

You're talking about the Tests and Release pipelines? If you find a way to do that, yes certainly, although perhaps in the future we'll want the test pipeline to go faster and the release pipeline to do more things, so if they can be kept modular that would be grand. :) 

> 2. There is some room for UX improvements: For example, we may want to add a --with-compiler flag to the wrapper executable to force the usage of a certain GHC compiler.

There is room indeed.

> 3. The only thing I can think of is a general suggestion to the user like Did you run cabal build --with-compiler VERSION ?.

That's a good thing, yes.

> 4. Since the wrapper/worker split of the executables is somewhat fragile we might want a testsuite checking that the print-api-VERSION gets picked.

Yes

> 5. We could extend the scope of it to be a downloader (i.e. download the appropriate bindist if a print-api-VERSION is not found) or an installer (i.e. download the source code, build and install the demanded print-api-VERSION if it is not found) or print instruction how to obtain it.

No I believe that is a job that is best left to ghcup. Otherwise a proper message that asks the user to get the correct binary will suffice.

> 6. Also, I think it was nice if the library provides the boiler plate to implement a Cabal testsuite checking API differences.

Do you have something clearer in mind? I could be interested. :)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,9 +1,7 @@
 name: Tests Pipeline
 
 on:
-  pull_request:
-  push:
-    branches: ['main']
+  pull_request: {}
```
Comment (tchoutri) [inline .github/workflows/ci.yml:4]: Why this change? GitHub is not helpful in pointing me to the commit that introduced it.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -27,64 +25,86 @@ jobs:
...
+
       - name: Cache
-        uses: actions/cache@v4.0.2
+        uses: actions/cache@v4
         with:
           path: ${{ steps.setup-haskell.outputs.cabal-store }}
-          key: ${{ runner.os }}-ghc-${{ matrix.ghc }}-${{ hashFiles('**/plan.json') }}
-          restore-keys: ${{ runner.os }}-ghc-${{ matrix.ghc }}-
+          key: ${{ runner.os }}-ghc-${{ steps.setup-haskell.outputs.ghc-version }}-${{ hashFiles('**/plan.json') }}
+          restore-keys: ${{ runner.os }}-ghc-${{ steps.setup-haskell.outputs.ghc-version }}-
```
Comment (tchoutri) [inline .github/workflows/ci.yml:58]: Why?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -127,43 +150,34 @@ jobs:
...
-      with:
-        name: artifact
-        path: ./out
-
-    - name: Release
-      uses: softprops/action-gh-release@v2
-      if: startsWith(github.ref, 'refs/tags/')
-      with:
-        draft: true
-        files: ./out/*
```
Comment (tchoutri) [inline .github/workflows/ci.yml:169]: Why do you remove the pre-release mechanism?

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -146,45 +165,81 @@ jobs:
...
 
-      - name: Upload print-api executable to workflow artifacts
+      - name: Upload the executables archive
         uses: actions/upload-artifact@v3
         with:
           name: artifact
           path: ${{ env.PRINTAPI_EXEC_TAR }}
 
+  # NOTE: We run upx on the executables in the tests/build-alpine job.
+  # That means, the assets that are linked dynamically _appear_ to be statically linked after that.
```
Comment (tchoutri) [inline .github/workflows/release.yml:201]: That's not quite the correct phrasing, I would rather say:
```suggestion
  # NOTE: We run upx on the executables in every linux job to compress them.
  # That means, the assets that are linked dynamically _appear_ to be statically linked after that, despite not being compiled statically in Alpine with musl. They are still dynamically-linked once they are decompresed.
```

**Review by tchoutri — CHANGES_REQUESTED** [review]

**Comment by Kleidukos** [comment]
> 1. I think some parts of the CI can be consolidated: Right now we have two workflow containing two jobs. Those four jobs are awfully similar, and we might just merge them (or some of them) into one (or two) jobs/workflows in order to reduce duplication.
> Would such a contribution be accepted?

You're talking about the Tests and Release pipelines? If you find a way to do that, yes certainly, although perhaps in the future we'll want the test pipeline to go faster and the release pipeline to do more things, so if they can be kept modular that would be grand. :) 

> 2. There is some room for UX improvements: For example, we may want to add a --with-compiler flag to the wrapper executable to force the usage of a certain GHC compiler.

There is room indeed.

> 3. The only thing I can think of is a general suggestion to the user like Did you run cabal build --with-compiler VERSION ?.

That's a good thing, yes.

> 4. Since the wrapper/worker split of the executables is somewhat fragile we might want a testsuite checking that the print-api-VERSION gets picked.

Yes

> 5. We could extend the scope of it to be a downloader (i.e. download the appropriate bindist if a print-api-VERSION is not found) or an installer (i.e. download the source code, build and install the demanded print-api-VERSION if it is not found) or print instruction how to obtain it.

No I believe that is a job that is best left to ghcup. Otherwise a proper message that asks the user to get the correct binary will suffice.

> 6. Also, I think it was nice if the library provides the boiler plate to implement a Cabal testsuite checking API differences.

Do you have something clearer in mind? I could be interested. :)

**Comment by Kleidukos** [comment]
@mmhat You removed the nightly release mechanism from the CI, which is not quite what I expected from this PR

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Kleidukos) [inline .github/workflows/ci.yml:4]: I intend to have the CI run on PRs that target the `main` branch since this is the workflow of this repository.

**Review by Kleidukos — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (mmhat) [reply .github/workflows/ci.yml]: Oh, sorry; I did not intend to change that workflow! Fixed.

**Comment by mmhat** [comment]
> @mmhat You removed the nightly release mechanism from the CI, which is not quite what I expected from this PR

... which is not quiet what I intended :upside_down_face: 

Just to clarify that we are talking about the same thing: By "nightly release" you mean the `print-api-head-*` artifacts uploaded during the `ci.yml` workflow?

**Comment by tchoutri** [comment]
Yes I do mean `print-api-head`. :)

**Comment by mmhat** [comment]
> You're talking about the Tests and Release pipelines? If you find a way to do that, yes certainly, although perhaps in the future we'll want the test pipeline to go faster and the release pipeline to do more things, so if they can be kept modular that would be grand. :)

I had nothing less in mind :slightly_smiling_face:
I think the (biggest) differences between the two workflows (currently in main) are:

1. Both run on pushes to main; release.yml if "v" tags are pushed and ci.yml runs for pull requests.
2. ci.yml has only a GHC 9.10.1 Alpine job; release.yml runs the whole matrix for Alpine.
3. Both have release jobs, but I believe the one in ci.yml never runs due to mutual excluding conditions. (I think I got this wrong in my previous comment.)

To me that looks like as if it could one workflow with just some run conditions for the release job and a conditional build matrix.

> No I believe that is a job that is best left to ghcup. Otherwise a proper message that asks the user to get the correct binary will suffice.

I totally; Just wanted to point out the options here.

> > 6. Also, I think it was nice if the library provides the boiler plate to implement a Cabal testsuite checking API differences.
> 
> Do you have something clearer in mind? I could be interested. :)

Nothing specific yet; Essentially I want a `defaultMain` that dumps the API of the current package and compares it to a golden value.

**Comment by Kleidukos** [comment]
> Just to clarify that we are talking about the same thing: By "nightly release" you mean the `print-api-head-*` artifacts uploaded during the `ci.yml` workflow?

Yes I am talking about the -head releases :)

**Comment by Kleidukos** [comment]
> Both have release jobs, but I believe the one in ci.yml never runs due to mutual excluding conditions.

Ah, it appears I messed up the `-head` pre-reload workflow then. This will have to be fixed after this PR lands.

**Comment by Kleidukos** [comment]
> Nothing specific yet; Essentially I want a defaultMain that dumps the API of the current package and compares it to a golden value.

Hmm, can't we do something that integrates with https://flora.pm/packages/@hackage/tasty-golden already?

File: install-for-ghcup-compilers.sh
Comment (Kleidukos) [inline install-for-ghcup-compilers.sh:1]: Two things: 

1. A message that says "We are going to install versions of `print-api` for the following versions of GHC" will be needed
2. We need to pipe the log/error output of ghcup to /dev/null, because otherwise we get:

```bash
❯ ghcup list --raw-format --tool ghc --show-criteria installed | cut -d ' ' -f 2

[ Warn  ] New cabal version available. If you want to install this latest version, run 'ghcup install cabal 3.12.1.0'
[ Warn  ] New stack version available. If you want to install this latest version, run 'ghcup install stack 3.1.1'
9.4.8
9.6.4
9.6.6
9.8.2
9.10.1
```

instead of:

```
❯ ghcup list --raw-format --tool ghc --show-criteria installed 2>/dev/null | cut -d ' ' -f 2

9.4.8
9.6.4
9.6.6
9.8.2
9.10.1
```

**Review by Kleidukos — CHANGES_REQUESTED** [review]

File: install-for-ghcup-compilers.sh
Comment (mmhat) [reply install-for-ghcup-compilers.sh:1]: Good idea; I'll change that during the next week.

**Comment by mmhat** [comment]
> > Both have release jobs, but I believe the one in ci.yml never runs due to mutual excluding conditions.
> 
> Ah, it appears I messed up the `-head` pre-reload workflow then. This will have to be fixed after this PR lands.

Can you open an issue with a short description how the workflows should work?  Maybe I'll find the time to fix that too.

Also, I noted that the workflows in the get-tested _seem_ to be identical (I just had a quick look, so I might be wrong). Should those be changed to?

**Comment by mmhat** [comment]
> > Nothing specific yet; Essentially I want a defaultMain that dumps the API of the current package and compares it to a golden value.
> 
> Hmm, can't we do something that integrates with https://flora.pm/packages/@hackage/tasty-golden already? 

Maybe; I don't know (yet). It was just an idea while I was working on this PR.

File: install-for-ghcup-compilers.sh
Comment (mmhat) [reply install-for-ghcup-compilers.sh:1]: Done.

**Comment by mmhat** [comment]
@Kleidukos I improved the `install-for-ghcup-compilers.sh` script as requested. Is there anything else I can do to push this PR over the line?

**Comment by Kleidukos** [comment]
@mmhat yes, two things:

1. There are unfortunately conflicts now, sorry.
2. If you could restore the pre-release mechanism that you removed, it would help with reviewing because the diff would be smaller. Let's discuss this in another ticket, as you suggested. :)

**Comment by mmhat** [comment]
> @mmhat yes, two things:
> 
> 1. There are unfortunately conflicts now, sorry.

No problem.

> 2. If you could restore the pre-release mechanism that you removed, it would help with reviewing because the diff would be smaller. Let's discuss this in another ticket, as you suggested. :)

I added the pre-release job as it is currently in the `main` branch. Does that meet your expectations or should I roll back more changes made in the CI?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -43,54 +41,69 @@ jobs:
...
+
       - name: Cache
-        uses: actions/cache@v4.0.2
+        uses: actions/cache@v4
         with:
           path: ${{ steps.setup-haskell.outputs.cabal-store }}
-          key: ${{ runner.os }}-ghc-${{ matrix.ghc }}-${{ hashFiles('**/plan.json') }}
-          restore-keys: ${{ runner.os }}-ghc-${{ matrix.ghc }}-
+          key: ${{ runner.os }}-ghc-${{ steps.setup-haskell.outputs.ghc-version }}-${{ hashFiles('**/plan.json') }}
+          restore-keys: ${{ runner.os }}-ghc-${{ steps.setup-haskell.outputs.ghc-version }}-
```
[the remark this answers is not in the document]
Comment (mmhat) [reply .github/workflows/ci.yml:74]: *TLDR;* It's a leftover and I can revert the change if you like.

Long version: As I wrote, at some point I wanted to know where the different jobs in the different pipelines were different.
So I extracted those into four files (more or less like `{ci,release}-{ubuntu,alpine}.yml`) and aimed for a minimal diff of those files. Since the Alpine jobs use `steps.setup-haskell...` and the Ubuntu jobs use `matrix.ghc...` I needed to choose, and I decided to go with the first one.
Since this change is not related to this PR I am happy to revert it if you like.

[the remark this answers is not in the document]
Comment (Kleidukos) [reply .github/workflows/ci.yml:74]: I'd be happy to consider these changes in their own PR at a later time indeed. Thank you for being understanding. :bow:

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -131,20 +146,23 @@ jobs:
...
 
       - name: Install
         run: |
-          bin=$(cabal -v0 --project-file=cabal.static.project list-bin print-api)
           mkdir distribution
-          cp ${bin} distribution/print-api
+          for executable in 'print-api' 'print-api-${{ matrix.ghc }}'; do
+            path=$(cabal -v0 --project-file=cabal.static.project list-bin "print-api:exe:${executable}")
+            cp "${path}" distribution
+          done
```
[the remark this answers is not in the document]
Comment (mmhat) [reply .github/workflows/ci.yml:165]: Every matrix job builds two executables: The `print-api` wrapper and the `print-api-<GHC VERSION>` executable.
The former one needs to be built and distributed by _some_ matrix job, but I am not sure how to decide on one.
Hence I decided to simply build and distribute it in all jobs.
An alternative solution would be to modify `get-tested` to provide more information:
For example, given `tested-with: GHC ==9.4.8, GHC ==9.10.1`, it could add a `newest` field in each matrix job indicating whether the job uses the newest GHC in the list:
```yaml
matrix:
  include:
    - ghc: "9.4.8"
      newest: false
    - ghc: "9.10.1"
      newest: true
```
Then we could use that information to built and distribute the `print-api` wrapper executable in job for the newest GHC only.

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -37,47 +36,55 @@ jobs:
...
+
       - name: Cache
-        uses: actions/cache@v4.0.2
+        uses: actions/cache@v4
         with:
           path: ${{ steps.setup-haskell.outputs.cabal-store }}
-          key: ${{ runner.os }}-ghc-${{ matrix.ghc }}-${{ hashFiles('**/plan.json') }}
-          restore-keys: ${{ runner.os }}-ghc-${{ matrix.ghc }}-
+          key: ${{ runner.os }}-ghc-${{ steps.setup-haskell.outputs.ghc-version }}-${{ hashFiles('**/plan.json') }}
+          restore-keys: ${{ runner.os }}-ghc-${{ steps.setup-haskell.outputs.ghc-version }}-
```
[the remark this answers is not in the document]
Comment (mmhat) [reply .github/workflows/release.yml:76]: See previous comment.

File: install-for-ghcup-compilers.sh
[the remark this answers is not in the document]
Comment (mmhat) [reply install-for-ghcup-compilers.sh:1]: Oh, good idea! I did that and applied the shellcheck suggestions.

File: .github/workflows/ci.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kleidukos) [reply .github/workflows/ci.yml:165]: Ok, let's merge this and see how to proceed later on.

File: .github/workflows/ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmhat) [reply .github/workflows/ci.yml]: Well, thank you for your patience!

**Comment by mmhat** [comment]
@Kleidukos The diff of the workflow files should be much smaller now.
Overall only the following changed in those:

1. I inlined the two scripts. This is the only change that is not necessary.
2. The post-process step for `ubuntu-latest` strips all executable.
3. The packaging step packs all executables.
4. The release job was rewritten so it merges all artifacts of the matrix builds.

**Review by Kleidukos — APPROVED** [review]
